webhooks.md

```diff
@@ -43,6 +43,199 @@ There are also additions/changes made to the `config` directory and to `main.go`
 
 It is furthermore recommended to add support for running the operator locally with webhooks.  The pattern for doing so can be seen here: https://github.com/openstack-k8s-operators/cinder-operator/pull/155/files.
 
+### Reusing webhook logic
+
+Both the defaulting and the validation logic of the webhook are implemented in
+the service operators but to apply them meaningfully we need to call the that
+logic form the openstack-operator as the webhhooks will run there in a real
+deployment. To do so we need to follow an API pattern described below.
+
+**Defaulting**
+
+For the example lets assume that the service CRD is called Foo
+
+In the service operator:
+```golang
+func (r *Foo) Default() {
+	r.Spec.Default()
+}
+
+func (spec *FooSpec) Default() {
+    // Implement the defaulting logic on the Spec struct here
+}
+```
+
+In the openstack-operator:
+```golang
+func (r *OpenStackControlPlane) Default() {
+	r.DefaultServices()
+}
+
+func (r *OpenStackControlPlane) DefaultServices() {
+	if r.Spec.Foo.Enabled {
+		r.Spec.Foo.Template.Default()
+	}
+```
+So when the webhook runs in the service operator the call chain is:
+1. Foo.Default()
+2. FooSpec.Default()
+
+When the webhook runs in the openstack-operator the call chain is:
+1. OpenStackControlPlane.Default()
+2. FooSpec.Default()
+
+Therefore the service operator should not implement any defaulting logic in
+Foo.Default but all should be in FooSpec.Default() or in deeper calls
+
+**Validation**
+
+The validation webhook is more complex than the defaulting one as it:
+1. Needs to return errors to the caller than can be different depending on
+   where the webhook runs
+2. Needs to specific the path of the field in the validated structure that
+   caused the validation error.
+3. Validation has multiple types, validation during create, update, or even
+   delete. The update validation has access to both the old and the new struct
+   value to be able to detect the change.
```

Comment:
I would focus on this a bit more. Specifically, it would be a good idea to elaborate on requirements, that is MUST and MUST NOT, of each validation type.

Reply:
I'm not sure what kind of requirements you are after. Do you have an example?

Reply:
Sure. I was thinking along the lines of:

> Create Validation MUST return error of the NewInvalid type, if it encounters any invalid section of the created data structure.

Reply:
Clarified it in the doc